src/thread_freezer.cpp

```diff
@@ -15,71 +15,112 @@ NtGetNextThread(HANDLE ProcessHandle, HANDLE ThreadHandle, ACCESS_MASK DesiredAc
 }
 
 namespace safetyhook {
-ThreadFreezer::ThreadFreezer() {
-    size_t num_threads_frozen{};
+void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint32_t, HANDLE, CONTEXT&)> visit_fn) {
+    // Freeze all threads.
+    int num_threads_frozen{};
 
     do {
-        num_threads_frozen = m_frozen_threads.size();
         HANDLE thread{};
+        num_threads_frozen = 0;
 
         while (true) {
+            HANDLE next_thread{};
             const auto status = NtGetNextThread(GetCurrentProcess(), thread,
                 THREAD_QUERY_LIMITED_INFORMATION | THREAD_SUSPEND_RESUME | THREAD_GET_CONTEXT | THREAD_SET_CONTEXT, 0,
-                0, &thread);
+                0, &next_thread);
+
+            if (thread != nullptr) {
+                CloseHandle(thread);
+            }
 
             if (!NT_SUCCESS(status)) {
                 break;
             }
 
+            thread = next_thread;
+
             const auto thread_id = GetThreadId(thread);
-            const auto already_frozen = std::any_of(m_frozen_threads.begin(), m_frozen_threads.end(),
-                [=](const auto& thread) { return thread.thread_id == thread_id; });
 
-            // Don't freeze ourselves or threads we already froze.
-            if (thread_id == 0 || thread_id == GetCurrentThreadId() || already_frozen) {
-                CloseHandle(thread);
+            if (thread_id == 0 || thread_id == GetCurrentThreadId()) {
+                continue;
+            }
+
+            const auto suspend_count = SuspendThread(thread);
+
+            if (suspend_count == static_cast<DWORD>(-1)) {
+                continue;
+            }
+
+            if (suspend_count != 0) { // Thread was already suspended.
```

Comment:
If this is using ResumeThread on already suspended threads here, might that mean the later thread resuming block would end up resuming the thread, even though it was already suspended outside of ThreadFreezer?

Might it be better to remove this block so that the suspend count would be incremented, so that the later resume block only decrements instead of actually resuming?

Reply:
I've modified the logic here to account for threads that were already frozen (f579cbf74e2d664bcaa354c07c2d2f290a148915). Give it a try and let me know how it works.

Reply:
Mostly seems good, tried with a few different games (CP2077/Witcher3/RDR2/TLOU) and by itself worked fine.

Also tested with the scenario I mentioned with all the mod tools, for CP2077/Witcher3/RDR2 it always seemed to work fine (which is a good sign since CP2077 did give ThreadFreezer issues before)

With TLOU it seems using SpecialK with safetyhook DLL has a rare chance of freezing on launch though, happened maybe 1 in 10 times I tried to launch, after removing either one of them I couldn't get freezing to happen again.
Odd thing is my safetyhook DLL was dynamically loaded as nvngx.dll, long after SpecialKs DllMain or games WinMain would have ran...
Maybe SK is doing something that delays hooking until later on, close to when nvngx.dll is loaded in maybe, not really sure how to look into that more though.

TLOU is a pretty notorious bad port though and apparently does weird things with CPU cores, maybe that's part of why this only seemed to happen with that.
(that did make me wonder if doing something with `SetProcessAffinityMask` before freezing could be worth looking into though?)